File: cratedb_sqlparse_py/cratedb_sqlparse/parser.py
from typing import List

from antlr4 import CommonTokenStream, InputStream, RecognitionException, Token
from antlr4.error.ErrorListener import ErrorListener

from cratedb_sqlparse.AstBuilder import AstBuilder
from cratedb_sqlparse.generated_parser.SqlBaseLexer import SqlBaseLexer
from cratedb_sqlparse.generated_parser.SqlBaseParser import SqlBaseParser
from cratedb_sqlparse.models import Metadata


def BEGIN_DOLLAR_QUOTED_STRING_action(self, localctx, actionIndex):
    if actionIndex == 0:
        self.tags.append(self.text)


def END_DOLLAR_QUOTED_STRING_action(self, localctx, actionIndex):
    if actionIndex == 1:
        self.tags.pop()


def END_DOLLAR_QUOTED_STRING_sempred(self, localctx, predIndex) -> bool:
    if predIndex == 0:
        return self.tags[0] == self.text
    return False


SqlBaseLexer.tags = []
SqlBaseLexer.BEGIN_DOLLAR_QUOTED_STRING_action = BEGIN_DOLLAR_QUOTED_STRING_action
SqlBaseLexer.END_DOLLAR_QUOTED_STRING_action = END_DOLLAR_QUOTED_STRING_action
SqlBaseLexer.END_DOLLAR_QUOTED_STRING_sempred = END_DOLLAR_QUOTED_STRING_sempred


class ParsingException(Exception):
    def __init__(self, *, query: str, msg: str, offending_token: Token, e: RecognitionException):
        self.message = msg
        self.offending_token = offending_token
        self.e = e
        self.query = query

    @property
    def error_message(self):
        return f"{self!r}[line {self.line}:{self.column} {self.message}]"

    @property
    def original_query_with_error_marked(self):
        query = self.offending_token.source[1].strdata
        offending_token_text: str = query[self.offending_token.start : self.offending_token.stop + 1]
        query_lines: list = query.split("\n")
        offending_line: str = query_lines[self.line - 1]

        # White spaces from the beginning of the offending line to the offending text, so the '^'
        # chars are correctly placed below the offending token.
        newline_offset = offending_line.index(offending_token_text)
        newline = (
            offending_line
            + "\n"
            + (" " * newline_offset + "^" * (self.offending_token.stop - self.offending_token.start + 1))
        )

        query_lines[self.line - 1] = newline

        msg = "\n".join(query_lines)
        return msg

    @property
    def column(self):
        return self.offending_token.column

    @property
    def line(self):
        return self.offending_token.line

    def __repr__(self):
        return type(self.e).__qualname__

    def __str__(self):
        return self.error_message


class CaseInsensitiveStream(InputStream):
    def LA(self, offset: int):
        result = super().LA(offset)
        if result <= 0 or result == Token.EOF:
            return result
        return ord(chr(result).upper())


class ExceptionErrorListener(ErrorListener):
    """
    Error listener that raises the antlr4 error as a Python exception.
    """

    errors = ()  # Dummy for passing tests, should not be used.

    def syntaxError(self, recognizer, offendingSymbol, line, column, msg, e):
        error = ParsingException(
            msg=msg,
            offending_token=offendingSymbol,
            e=e,
            query=e.ctx.parser.getTokenStream().getText(e.ctx.start, e.offendingToken.tokenIndex),
        )
        raise error


class ExceptionCollectorListener(ErrorListener):
    """
    Error listener that collects all errors into errors for further processing.

    Based partially on https://github.com/antlr/antlr4/issues/396
    """

    def __init__(self):
        self.errors = []

    def syntaxError(self, recognizer, offendingSymbol, line, column, msg, e):
        error = ParsingException(
            msg=msg,
            offending_token=offendingSymbol,
            e=e,
            query=e.ctx.parser.getTokenStream().getText(e.ctx.start, e.offendingToken.tokenIndex),
        )

        self.errors.append(error)


class Statement:
    """
    Represents a CrateDB SQL statement.
    """

    def __init__(self, ctx: SqlBaseParser.StatementContext, exception: ParsingException = None):
        self.ctx: SqlBaseParser.StatementContext = ctx
        self.exception = exception
        self.metadata = Metadata()

    @property
    def tree(self):
        """
        Returns the String representation of the Tree in LISP format.
        """
        return self.ctx.toStringTree(recog=self.ctx.parser)

    @property
    def original_query(self):
        """
        The original query that was fed into `sqlparse`, including semicolons and comments.
        """
        return self.ctx.parser.getTokenStream().getText()

    @property
    def query(self) -> str:
        """
        Returns the query, comments and ';' are not included.
        """
        return self.ctx.parser.getTokenStream().getText(start=self.ctx.start, stop=self.ctx.stop)

    @property
    def type(self):
        """
        Returns the type of the Statement, i.e.: 'SELECT, 'UPDATE', 'COPY TO'...
        """
        return self.ctx.start.text

    def __repr__(self):
        return f'{self.__class__.__qualname__}<{self.query if len(self.query) < 15 else self.query[:15] + "..."}>'


def find_suitable_error(statement, errors):
    for error in errors[:]:
        # We clean the error_query of ';' and spaces because ironically,
        # we can get the full query in the error but not in the parsed statement.
        error_query = error.query
        if error_query.endswith(";"):
            error_query = error_query[: len(error_query) - 1]

        error_query = error_query.lstrip().rstrip()

        # If a good match error_query contains statement.query
        if statement.query in error_query:
            statement.exception = error
            errors.pop(errors.index(error))


def sqlparse(query: str, raise_exception: bool = False) -> List[Statement]:
    """
    Parses a string into SQL `Statement`.
    """
    input_ = CaseInsensitiveStream(query)
    lexer = SqlBaseLexer(input_)
    lexer.removeErrorListeners()
    stream = CommonTokenStream(lexer)

    parser = SqlBaseParser(stream)
    parser.removeErrorListeners()
    error_listener = ExceptionErrorListener() if raise_exception else ExceptionCollectorListener()
    parser.addErrorListener(error_listener)

    tree = parser.statements()

    statements_context: list[SqlBaseParser.StatementContext] = list(
        filter(lambda children: isinstance(children, SqlBaseParser.StatementContext), tree.children)
    )

    statements = []

    for statement_context in statements_context:
        stmt = Statement(statement_context)
        if len(statements_context) == 1 and error_listener.errors:
            # There is only one statement parsed, no need to get fancy to match.
            stmt.exception = error_listener.errors.pop()

        else:
            find_suitable_error(stmt, error_listener.errors)

        statements.append(stmt)

    else:
        # We might still have error(s) that we couldn't match with their origin statement,
        # this happens when the query is composed of only one keyword, e.g. 'SELCT 1'
        # the error.query will be 'SELCT' instead of 'SELCT 1'.
        if len(error_listener.errors) == 1:
            # This case has an edge case where we hypothetically assign the
            # wrong error to a statement, for example,
            #     SELECT A FROM tbl1;
            #     SELEC 1;
            #          ^
            # This would match both conditionals, this however, is protected
            # by https://github.com/crate/cratedb-sqlparse/issues/28, but might
            # change in the future.
            error = error_listener.errors[0]
            for stmt in statements:
                if stmt.exception is None and error.query in stmt.query:
                    stmt.exception = error
                    break

        if len(error_listener.errors) > 1:
            # We might have too much ambiguity to match errors, but also one statement can
            # generate several exceptions, and we only support one per statement so this could be
            # triggered and not be an actual error.
            pass

    # We extract the metadata and enrich every Statement's `metadata`.
    stmt_enricher = AstBuilder()

    for stmt in statements:
        stmt_enricher.enrich(stmt)

    return statements

File: cratedb_sqlparse_js/cratedb_sqlparse/parser.js
import SqlBaseLexer from "./generated_parser/SqlBaseLexer.js";
import SqlBaseParser from "./generated_parser/SqlBaseParser.js";
import {CommonTokenStream, ErrorListener, InputStream, Interval, Token} from "antlr4";
import {AstBuilder} from "./AstBuilder.js";
import {Metadata} from "./models.js"
function BEGIN_DOLLAR_QUOTED_STRING_action(localctx, actionIndex) {
    if (actionIndex === 0) {
        this.tags.push(this.text);
    }
}

function END_DOLLAR_QUOTED_STRING_action(localctx, actionIndex) {
    if (actionIndex === 1) {
        this.tags.pop();
    }
}

function END_DOLLAR_QUOTED_STRING_sempred(localctx, predIndex) {
    if (predIndex === 0) {
        return this.tags[0] === this.text;
    }
}

SqlBaseLexer.prototype.tags = [];
SqlBaseLexer.prototype.BEGIN_DOLLAR_QUOTED_STRING_action = BEGIN_DOLLAR_QUOTED_STRING_action;
SqlBaseLexer.prototype.END_DOLLAR_QUOTED_STRING_action = END_DOLLAR_QUOTED_STRING_action;
SqlBaseLexer.prototype.END_DOLLAR_QUOTED_STRING_sempred = END_DOLLAR_QUOTED_STRING_sempred;

export class ParseError extends Error {
    name = 'ParseError'

    /**
     *
     * @param {string} query
     * @param {string} msg
     * @param {object} offending_token
     * @param {object} e
     * @member {string} errorMessage
     * @member {string} errorMessageVerbose
     */
    constructor(query, msg, offending_token, e) {
        super(msg);
        this.query = query;
        this.msg = msg;
        this.offendingToken = offending_token;
        this.line = this.getLine();
        this.column = this.getColumn();
        this.errorMessage = this._getErrorMessage();
        this.errorMessageVerbose = this.getOriginalQueryWithErrorMarked()
    }

    _getErrorMessage() {
        return `[line ${this.line}:${this.column} ${this.message}]`
    }

    /**
     *
     * @returns {Number}
     */
    getColumn() {
        return this.offendingToken.column
    }

    /**
     *
     * @returns {Number}
     */
    getLine() {
        return this.offendingToken.line
    }

    /**
     *
     * @returns {string}
     */
    getOriginalQueryWithErrorMarked() {
        const query = this.offendingToken.source[1].strdata
        const offendingTokenText = query.substring(this.offendingToken.start, this.offendingToken.stop + 1)
        const queryLines = query.split("\n")
        const offendingLine = queryLines[this.getLine() - 1]
        const newLineOffset = offendingLine.indexOf(offendingTokenText)

        const newline = (
            offendingLine
            + "\n"
            + (" ".repeat(newLineOffset) + "^".repeat(this.offendingToken.stop - this.offendingToken.start + 1))
        )
        queryLines[this.line - 1] = newline

        return queryLines.join("\n")
    }
}

class CaseInsensitiveStream extends InputStream {
    LA(offset) {
        const result = super.LA(offset);
        if (result <= 0 || result === Token.EOF) {
            return result;
        }
        return String.fromCharCode(result).toUpperCase().charCodeAt(0);
    }
}

class ExceptionErrorListener extends ErrorListener {
    errors = []
    syntaxError(recognizer, offendingSymbol, line, column, msg, e) {
        throw new ParseError(
            e.ctx.parser.getTokenStream().getText(new Interval(
                e.ctx.start,
                e.offendingToken.tokenIndex)
            ),
            msg,
            offendingSymbol,
            e
        )
    }
}

class ExceptionCollectorListener extends ErrorListener {
    constructor() {
        super();
        this.errors = [];
    }

    syntaxError(recognizer, offendingSymbol, line, column, msg, e) {
        super.syntaxError(recognizer, offendingSymbol, line, column, msg, e);
        const error = new ParseError(
            e.ctx.parser.getTokenStream().getText(new Interval(
                e.ctx.start,
                e.offendingToken.tokenIndex)
            ),
            msg,
            offendingSymbol,
            e
        )
        this.errors.push(error)
    }
}


/*
* Represents a CrateDB SQL statement.
* */
export class Statement {

    /**
     *
     * @member {Statement} query
     * @member {string} originalQuery
     * @member {Metadata} metadata
     * @member {string} type - The type of query, example: 'SELECT'
     * @member {string} tree
     * @param {object} ctx
     * @param {ParseError} exception
     */
    constructor(ctx, exception) {
        this.query = ctx.parser.getTokenStream().getText(
            new Interval(
                ctx.start.tokenIndex,
                ctx.stop.tokenIndex,
            )
        )
        this.originalQuery = ctx.parser.getTokenStream().getText();
        this.tree = ctx.toStringTree(null, ctx.parser);
        this.type = ctx.start.text;
        this.ctx = ctx;
        this.exception = exception || null;
        this.metadata = new Metadata();
    }
}

/**
 *
 * @param {string} string
 * @returns {string}
 */
function trim(string) {
    return string.replace(/^\s+|\s+$/gm, '');
}


function findSuitableError(statement, errors) {
    for (const error of errors) {
        let errorQuery = error.query;

        if (errorQuery.endsWith(";")) {
            errorQuery = errorQuery.substring(0, errorQuery.length - 1);
        }

        errorQuery = trim(errorQuery);

        // If a good match error_query contains statement.query
        if (errorQuery.includes(statement.query)) {
            statement.exception = error;
            errors.splice(errors.indexOf(error), 1);
        }
    }
}

/**
 *
 * @param {string} query
 * @param {Boolean} raise_exception
 * @returns {Statement[]}
 */
export function sqlparse(query, raise_exception = false) {
    const input = new CaseInsensitiveStream(query);
    const lexer = new SqlBaseLexer(input);
    lexer.removeErrorListeners();
    const stream = new CommonTokenStream(lexer);

    const parser = new SqlBaseParser(stream);
    parser.removeErrorListeners();

    const errorListener = raise_exception ? new ExceptionErrorListener() : new ExceptionCollectorListener()

    parser.addErrorListener(errorListener);

    const tree = parser.statements();

    const statementsContext = tree.children.filter((children) => children instanceof SqlBaseParser.StatementContext)

    let statements = []
    for (const statementContext of statementsContext) {
        let stmt = new Statement(statementContext)

        if (statementsContext.length === 1 && errorListener.errors) {
            stmt.exception = errorListener.errors.pop();

        } else {
            findSuitableError(stmt, errorListener.errors)
        }

        statements.push(stmt)
    }

    if (errorListener.errors.length === 1) {
        // Fixme, what if there are two unassigned errors ?
        // can that even be possible?
        let error = errorListener.errors[0]

        for (const stmt of statements) {
            if (stmt.exception === null && stmt.query.includes(error.query)) {
                stmt.exception = error
                break;
            }
        }
    }

    if (errorListener.errors.length > 1) {
        console.error("Could not match errors to queries, too much ambiguity, please report it opening an issue with the query.")
    }


    const stmtEnricher = new AstBuilder()

    for (const stmt of statements) {
        stmtEnricher.enrich(stmt)
    }

    return statements
}